File: tests/robot-java/src/main/java/frc/robot/subsystems/Elevator.java
package frc.robot.subsystems;

import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.RobotController;
import edu.wpi.first.wpilibj.simulation.ElevatorSim;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

public class Elevator extends SubsystemBase {
  private static final double kP = 4;
  private static final double kI = 0.0;
  private static final double kD = 0.0;

  private static final double kElevatorGearing = 10.0;
  private static final double kElevatorDrumRadius = Units.inchesToMeters(2);
  private static final double kCarriageMass = 4.0;

  private static final DCMotor kElevatorGearbox = DCMotor.getVex775Pro(4);
  private static final double kMinElevatorHeight = Units.inchesToMeters(0);
  private static final double kMaxElevatorHeight = Units.inchesToMeters(50);

  private final WPI_TalonSRX m_motor;
  private final PIDController m_controller;

  private double m_setpoint;

  // Sim
  private ElevatorSim m_elevatorSim;

  /** Create a new elevator subsystem. */
  @SuppressWarnings("this-escape")
  public Elevator() {
    m_controller = new PIDController(kP, kI, kD);
    m_motor = new WPI_TalonSRX(PortMap.kElevatorMotorPort);

    m_controller.setTolerance(0.005);

    if (RobotBase.isSimulation()) {
      m_elevatorSim =
          new ElevatorSim(
              kElevatorGearbox,
              kElevatorGearing,
              kCarriageMass,
              kElevatorDrumRadius,
              kMinElevatorHeight,
              kMaxElevatorHeight,
              true,
              0);
    }
  }

  public void log() {
    SmartDashboard.putNumber("Elevator Height", m_motor.getSelectedSensorPosition());
  }

  public void setVoltage(double output) {
    m_motor.set(output);
  }

  public void goToHeight(double height) {
    double pidVoltage = m_controller.calculate(m_motor.getSelectedSensorPosition(), height);
    setVoltage(pidVoltage);
  }

  public boolean isAtHeight() {
    double error = m_setpoint - m_motor.getSelectedSensorPosition();
    return error < 0.5;
  }

  @Override
  public void periodic() {
    log();
  }

  @Override
  public void simulationPeriodic() {
    m_elevatorSim.setInput(m_motor.get() * RobotController.getInputVoltage());
    m_elevatorSim.update(0.02);
    m_motor.getSimCollection().setQuadratureRawPosition((int) m_elevatorSim.getPositionMeters());
  }

  public void stop() {
    m_motor.set(0);
  }
}

File: tests/robot-cpp/subsystems/elevator.cpp
#include "robot-cpp/subsystems/elevator.hpp"

#include <frc/RobotController.h>
#include <frc/controller/PIDController.h>
#include <frc/livewindow/LiveWindow.h>
#include <frc/smartdashboard/SmartDashboard.h>

#include "robot-cpp/subsystems/ports.hpp"

namespace {
constexpr double kP = 5.0;
constexpr double kI = 0.0;
constexpr double kD = 0.0;
constexpr units::volt_t kGravityOffset{0.85};

constexpr double kElevatorGearing = 10.0;
constexpr units::meter_t kElevatorDrumRadius = 2_in;
constexpr units::kilogram_t kCarriageMass = 4.0_kg;

constexpr units::meter_t kMinElevatorHeight = 0_in;
constexpr units::meter_t kMaxElevatorHeight = 50_in;

frc::DCMotor kElevatorGearbox = frc::DCMotor::Vex775Pro(4);
}  // namespace

Elevator::Elevator()
    : m_motor{kElevatorMotorPort},
      m_controller{kP, kI, kD},
      m_elevatorSim(kElevatorGearbox, kElevatorGearing, kCarriageMass,
                    kElevatorDrumRadius, kMinElevatorHeight, kMaxElevatorHeight,
                    true, units::meter_t{0}) {
  m_controller.SetTolerance(0.005);
}

void Elevator::Log() {
  frc::SmartDashboard::PutNumber("Elevator Height (m)",
                                 GetElevatorHeight().to<double>());
}

units::meter_t Elevator::GetElevatorHeight() {
  return units::meter_t{m_motor.GetSelectedSensorPosition()};
}

void Elevator::SetVoltage(double output) {
  m_motor.SetVoltage(kGravityOffset + units::volt_t(output));
}

void Elevator::GoToHeight(units::meter_t height) {
  m_setpoint = height;

  double pidVoltage = m_controller.Calculate(GetElevatorHeight().to<double>(),
                                             height.to<double>());
  SetVoltage(pidVoltage);
}

bool Elevator::IsAtHeight() {
  units::meter_t error = m_setpoint - GetElevatorHeight();
  return error < 2_in;
}

void Elevator::Periodic() { Log(); }

void Elevator::SimulationPeriodic() {
  m_elevatorSim.SetInput(Eigen::Vector<double, 1>(
      m_motor.Get() * frc::RobotController::GetInputVoltage()));
  m_elevatorSim.Update(20_ms);
  m_motor.GetSimCollection().SetQuadratureRawPosition(
      m_elevatorSim.GetPosition().to<double>());
}

void Elevator::Stop() { m_motor.Set(0); }
